### PR TITLE
Sphinx - Leidy Martinez & Aleida Vieyra

### PR DESCRIPTION
**Pull request by Leidy-Martinez**

### CODE REVIEW

File: viewing_party/party.py [hunk 1]
```diff
@@ -1,23 +1,194 @@
...
+    movie= {}
+
+    if movie_title and genre and rating:
+    
+        movie["title"] = movie_title
+        movie["genre"] = genre
+        movie["rating"] = rating          
+        return movie
+
+    return None
```
Comment (mikellewade) [inline viewing_party/party.py:15]: One way to shorten this could be like the following: 
```python
def create_movie(movie_title, genre, rating):

    if not movie_title or not genre or not rating:
        return None
	
	return {
		"title": movie_title,
		"genre": genre,
		"rating": rating
	}
```
I would be okay with this since all we are doing is creating a dictionary representation of a movie and then return it.

File: viewing_party/party.py [hunk 2]
```diff
@@ -1,23 +1,194 @@
...
+
+def add_to_watched(user_data, movie):
+
+    # Create a copy of user data to avoid modifying original
+    updated_user_data = {
+            "watched": user_data["watched"][:]}
```
Comment (mikellewade) [inline viewing_party/party.py:21]: I see that you took into consideration that `user_data` was a mutable data type and decided to make a copy of the structure to avoid modifications of the original. I love that you are thinking about things like and applying your learnings from class! In this case, its actually fine to modify the original `user_data` since they ask you too in the requirements. Since the updated structure is returned and that return value is what is tested all your tests pass. If we were to test against the original `user_data` it would not.

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,194 @@
...
+    updated_user_data = {
+            "watchlist": user_data["watchlist"][:],
+            "watched": user_data["watched"][:]
+        } 
+    
+    # Check if movie with the title is in watchlist 
+    # and add it to the watched list
+    for movie in updated_user_data["watchlist"]:
+        if movie["title"] == title:
+            updated_user_data["watchlist"].remove(movie)
```
Comment (mikellewade) [inline viewing_party/party.py:49]: So now that we have talked about Big O analysis, we know that `.remove` has a time complexity of `O(n)` since it needs to look through every element in a list to find the sought after value. Right now, this loop has an outer loop of  `O(n)` and you when you find the movie to remove you have another `O(n)` operation. How could we circumvent using `remove`? (Hint: Think about list comprehension or ~~`pop` method~~) 

Also in this implementation, we will only use `remove` once, but in other filtering problems we might have to do this for multiple elements so it doesn't hurt to get in the habit now.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,194 @@
...
+    count = 0.0
+    sum = 0.0
+
+    if len(user_data["watched"]) == 0:
+        return sum
+    
+    # Count and sum rating values
+    for movie in user_data["watched"]:
+        if movie["rating"]:
+            count += 1
```
Comment (mikellewade) [inline viewing_party/party.py:68]: Do we need to keep count of the number of movies a user has watched or can we just use a built in function provided to us by python instead in this function? 👀

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,194 @@
...
+        if movie["genre"] in genre_frequencies:
+            genre_frequencies[movie["genre"]] += 1
+        else:
+            genre_frequencies[movie["genre"]] = 1
+
+    # Check for the genre with highest frequencies   
+    for key, current_value in genre_frequencies.items():
+        if current_value > max_value:
+            max_value = current_value
+            most_watched = key #retrieve the key with the highest value
```
Comment (mikellewade) [inline viewing_party/party.py:96]: Yessss, great job! ⭐️

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,194 @@
...
+
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    # Create key,value pairs {"genre" : genre_frequencies} in dict.
+    for movie in  user_data["watched"]:
+        if movie["genre"] in genre_frequencies:
+            genre_frequencies[movie["genre"]] += 1
+        else:
+            genre_frequencies[movie["genre"]] = 1
```
Comment (mikellewade) [inline viewing_party/party.py:90]: How could we use the `.get` method so that we could condense this to a single line? 👀

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,194 @@
...
 # -----------------------------------------
+def get_unique_watched(user_data):
+    # ONLY THE USER HAS WATCHED, BUT THE FRIENDS HAVE NOT
+    user_unique_watched = []
+    friends_watched = []
+    
+    # Get every movie from friends and add it to a new list
+    for friend in user_data["friends"]:
+        for movie in friend['watched']:
+            friends_watched.append(movie)
```
Comment (mikellewade) [inline viewing_party/party.py:112]: Love the flattening of your data structures so you can simplify your loops! Also what if we just added the title of movies to `friends_watched` instead of the whole movie?

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,194 @@
...
+    
+    # Get every movie from user and add it to a list
+    # only if is not in friends_wacthed list
+    for movie in user_data["watched"]:
+        if movie not in friends_watched:
```
Comment (mikellewade) [inline viewing_party/party.py:117]: Since `friends_watched` is a list, this line has a time complexity of `O(n)` (`n` being the number of movies in a list). What other structure could we use in order to circumvent the linear time complexity and get a constant one?

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,194 @@
...
+            user_unique_watched.append(movie)
+
+    return user_unique_watched
+
+
+def get_friends_unique_watched(user_data):
+    # AT LEAST ONE OF THE FRIENDS HAS WATCHED,
+    # BUT THE USER HAS NOT
+
+    user_watched = [movie for movie in user_data["watched"]]
```
Comment (mikellewade) [inline viewing_party/party.py:127]: Omg, list comprehension! Love to see her! 😍

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,194 @@
...
+    friends_unique_watched = []
+
+    # Get every movie from friends and add it to a new list:
+    # only if is not in the movies user has watched
+    # AND if it's not already in the list
+    for friend in user_data["friends"]:
+        for movie in friend['watched']:
+            if movie not in user_watched\
+            and movie not in friends_unique_watched:
```
Comment (mikellewade) [inline viewing_party/party.py:136]: ⭐️

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,194 @@
...
+    # List of movies that at least one friend has watched and user hasn't
+    friends_uniques = get_friends_unique_watched(user_data)
+    
+    # From friends movies add the movie to the list only if 
+    # the host is in user subscriptions
+    for movie in friends_uniques:
+        if movie["host"] in user_data["subscriptions"]:
+            rec_list.append(movie)
+            
+    return rec_list
```
Comment (mikellewade) [inline viewing_party/party.py:159]: Great and concise function! 😌

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,194 @@
...
+    user_freq_genre = get_most_watched_genre(user_data)
+    friends_unique_movies = get_friends_unique_watched(user_data)
+    
+
+    rec_list = []
+    
+    # Check the movies in friends that correspond to the same genre
+    # the user watch the most and add it to a list
+    for movie in friends_unique_movies:
+        if user_freq_genre in movie.values():
```
Comment (mikellewade) [inline viewing_party/party.py:176]: Would it be more declarative to say: 

```python
if movie["genre"] == user_freq_genre 
```
This more directly communicates that you are only adding movies that match the genre of a user's favorite genre

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,194 @@
...
+
+    rec_list = []
+    
+    # Check user unique movies and add them to a list if
+    # were marked as favorites
+    for movie in user_only_movies:
+        if movie in user_data["favorites"]:
+            rec_list.append(movie)
+            
+    return rec_list
```
Comment (mikellewade) [inline viewing_party/party.py:194]: Awesome function, great work! ⭐️

File: tests/test_wave_05.py [hunk 14]
```diff
@@ -52,13 +52,18 @@ def test_new_genre_rec_from_empty_friends():
...
+    
+    #Act
+    genre_recommendations = get_new_rec_by_genre(sonyas_data)
 
-    raise Exception("Test needs to be completed.")
+    #Assert
+    #raise Exception("Test needs to be completed.")
     # *********************************************************************
     # ****** Complete the Act and Assert Portions of these tests **********
+    assert len(genre_recommendations) == 0
```
Comment (mikellewade) [inline tests/test_wave_05.py:63]: What if we did `assert len(genre_recommendations) == []`? This more readily declares that the `genre_recommendations` should be an empty list. Also for testing, if someone returned, say an empty string, this test would technically pass.

File: tests/test_wave_03.py [hunk 15]
```diff
@@ -55,12 +55,14 @@ def test_friends_unique_movies_not_duplicated():
     # Assert
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
+    assert friends_unique_movies == [FANTASY_4, HORROR_1, INTRIGUE_3]
+    assert amandas_data["watched"] not in friends_unique_movies
```
Comment (mikellewade) [inline tests/test_wave_03.py:62]: So, this assert statement here is checking if the **entire** list of watched movies is **a single element** inside of `friends_unique_movies`. How could we change this to check if the individual movies are inside of `friends_unique_watched`?

File: tests/test_wave_01.py [hunk 16]
```diff
@@ -159,12 +159,13 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
```
Comment (mikellewade) [inline tests/test_wave_01.py:165]: ⭐️

File: tests/test_wave_01.py [hunk 17]
```diff
@@ -183,12 +184,17 @@ def test_moves_movie_from_watchlist_to_watched():
...
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
+    janes_data = {
+        "watchlist": [FANTASY_1],
+        "watched": [FANTASY_2, movie_to_watch]
+    }
+    assert updated_data == janes_data
```
Comment (mikellewade) [inline tests/test_wave_01.py:194]: Only suggestion I have here is to change name to something else like `expected`. Also see my note below about [making copies of the originals](https://github.com/Ada-C22/viewing-party/pull/14/files#r1792645935).
